src/nbl/video/CVulkanLogicalDevice.cpp

```diff
@@ -1425,6 +1425,118 @@ void CVulkanLogicalDevice::createGraphicsPipelines_impl(
         std::fill_n(output,vk_createInfos.size(),nullptr);
 }
 
+void CVulkanLogicalDevice::createRayTracingPipelines_impl(
+    IGPUPipelineCache* const pipelineCache,
+    const std::span<const IGPURayTracingPipeline::SCreationParams> createInfos,
+    core::smart_refctd_ptr<IGPURayTracingPipeline>* const output,
+    const IGPURayTracingPipeline::SCreationParams::SSpecializationValidationResult& validation
+)
+{
+    using SShaderGroupParams = asset::IRayTracingPipelineBase::SShaderGroupsParams;
+    using SGeneralShaderGroup = asset::IRayTracingPipelineBase::SGeneralShaderGroup;
+    using SHitShaderGroup = asset::IRayTracingPipelineBase::SHitShaderGroup;
+
+    const VkPipelineCache vk_pipelineCache = pipelineCache ? static_cast<const CVulkanPipelineCache*>(pipelineCache)->getInternalObject():VK_NULL_HANDLE;
+    
+    size_t maxShaderStages = 0;
+    for (const auto& info : createInfos)
+        maxShaderStages += info.shaders.size();
+    core::vector<VkRayTracingPipelineCreateInfoKHR> vk_createInfos(createInfos.size(), { VK_STRUCTURE_TYPE_RAY_TRACING_PIPELINE_CREATE_INFO_KHR,nullptr });
+    core::vector<VkPipelineShaderStageRequiredSubgroupSizeCreateInfo> vk_requiredSubgroupSize(maxShaderStages,{
+        VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_REQUIRED_SUBGROUP_SIZE_CREATE_INFO,nullptr
+    });
+    core::vector<VkPipelineShaderStageCreateInfo> vk_shaderStage(maxShaderStages, { VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO, nullptr });
+    core::vector<VkRayTracingShaderGroupCreateInfoKHR> vk_shaderGroup(maxShaderStages, { VK_STRUCTURE_TYPE_RAY_TRACING_SHADER_GROUP_CREATE_INFO_KHR, nullptr});
```

Comment:
you can reference the same shader multiple times across multiple hit groups, you need to come up with the upper bound on the size of `vk_shaderGroup`